cli/src/main/java/com/devonfw/tools/ide/commandlet/FileExtractor/CommandletFileExtractorImpl.java

```diff
@@ -0,0 +1,157 @@
+      this.context.trace("Determined file extension {}", extension);
+      TarCompression tarCompression = TarCompression.of(extension);
+      ExtractorFileType fileType = ExtractorFileType.valueOf(extension.toUpperCase());
+
+      if (tarCompression != null) {
+        fileAccess.untar(file, tmpDir, tarCompression);
+      } else if (fileType == ExtractorFileType.ZIP || fileType == ExtractorFileType.JAR) {
+        fileAccess.unzip(file, tmpDir);
+      } else if (fileType == ExtractorFileType.DMG) {
+        extractLinuxDmg(file, tmpDir);
+      } else if (fileType == ExtractorFileType.MSI) {
+        extractWindowsMsi(file, tmpDir);
+      } else if (fileType == ExtractorFileType.PKG) {
+        extractPkg(file, tmpDir);
+
+      } else {
+        throw new IllegalStateException("Unknown archive format " + extension + ". Can not extract " + file);
+      }
+      moveAndProcessExtraction(getProperInstallationSubDirOf(tmpDir), targetDir);
+      fileAccess.delete(tmpDir);
+    } else {
+      this.context.trace("Extraction is disabled for '{}' hence just moving the downloaded file {}.",
+          commandlet.getName(), file);
+
+      if (Files.isDirectory(file)) {
+        fileAccess.move(file, targetDir);
+      } else {
+        try {
+          Files.createDirectories(targetDir);
+        } catch (IOException e) {
+          throw new IllegalStateException("Failed to create folder " + targetDir);
+        }
+        fileAccess.move(file, targetDir.resolve(file.getFileName()));
+      }
+    }
+  }
+
+  @Override
+  public void moveAndProcessExtraction(Path from, Path to) {
+
+    this.context.getFileAccess().move(from, to);
+  }
+
+  /**
+   * @param path the {@link Path} to start the recursive search from.
+   * @return the deepest subdir {@code s} of the passed path such that all directories between {@code s} and the passed
+   *         path (including {@code s}) are the sole item in their respective directory and {@code s} is not named
+   *         "bin".
+   */
+  private Path getProperInstallationSubDirOf(Path path) {
+
+    try (Stream<Path> stream = Files.list(path)) {
+      Path[] subFiles = stream.toArray(Path[]::new);
+      if (subFiles.length == 0) {
+        throw new CliException("The downloaded package for the tool " + commandlet.getName()
+            + " seems to be empty as you can check in the extracted folder " + path);
+      } else if (subFiles.length == 1) {
+        String filename = subFiles[0].getFileName().toString();
+        if (!filename.equals(IdeContext.FOLDER_BIN) && !filename.equals(IdeContext.FOLDER_CONTENTS)
+            && !filename.endsWith(".app") && Files.isDirectory(subFiles[0])) {
+          return getProperInstallationSubDirOf(subFiles[0]);
+        }
+      }
+      return path;
+    } catch (IOException e) {
+      throw new IllegalStateException("Failed to get sub-files of " + path);
+    }
+  }
+
+  private void extractLinuxDmg(Path file, Path tmpDir) {
```

Comment:
`extractLinuxDmg` is the incorrect name as `DMG` files are specific for MacOS.
IMHO I would strip out `Windows` and `Linux` from these methods and instead add such hints to the JavaDoc of the methods in the interface making all these methods public.